src/dune_engine/build_system.ml

```diff
@@ -2291,12 +2296,27 @@ let run ?(report_error = Dune_util.Report_error.report) f =
   in
   let f () =
     let* () = t.handler.build_event Start in
+    let build_end_reported = ref false in
     let* res =
       Fiber.with_error_handler ~on_error:reraise_exn (fun () ->
           Memo.Build.run_with_error_handler (f ())
-            ~handle_error_no_raise:(report_early_exn ~report_error))
+            ~handle_error_no_raise:(fun exn ->
+              Fiber.fork_and_join_unit
+                (fun () -> report_early_exn ~report_error exn)
+                (fun () ->
```

Comment:
This looks like a weird place to report cancellation. First of all, the guarantee that some errors will be emitted after cancellation seems like a weird guarantee (it may be true now, but it seems more efficient and more reasonable to never emit events after cancellation).
Second, here you have to deal with duplicate events, and it's unclear why that's useful.

Why is the already-existing notification `t.handler t.config Build_interrupted` in `let rec iter` not sufficient?